### DISCUSSION TITLE
Feature suggestion: new nullability attribute: NotNullWhenMethodReturnsSuccessfully

### DISCUSSION THREAD

**Discussion by johnknoop** [dotnet/csharplang#4127] [Language Ideas] [answered]
A lot of unit test assertion libraries have methods such as:

```cs
void AssertNotNull(string? variableToTest)
{
    // will throw if variableToTest is null
}
```

But unfortunately there's no way for these libraries to let the static flow analysis know that it won't be null if the method returns without throwing, since no such attribute exists.

Here's an example using FluentAssertions, one of the most popular assertion libraries for C#:

![image](https://user-images.githubusercontent.com/3394270/98859123-f7a2c780-2461-11eb-9baa-a73391cfa15a.png)

Would it be possible to add a new nullability attribute that tells the compiler that an argument will not be null, if the method returns successfully?

Maybe something like `[NotNullWhenMethodReturnsSuccessfully]` and `[NullWhenMethodReturnsSuccessfully]` (you can probably come up with better names).

**Comment by 333fred** [comment, marked as answer]
Well, the original example could just be defined as:
```cs
void AssertNotNull([NotNull] string? variableToTest)
{
    // will throw if variableToTest is null
}
```
This would provide exactly the semantics you want.

The second example, though, I'm not sure is possible. You'd need to have the results of a second method somehow passing through inference to the parameter of the first method, and I'm not really sure there's a way to do this. Maybe a particularly specialized attribute could do this, but I don't know how that could be generally-applicable enough to be worth adding to the language.

**Reply by johnknoop** [reply]
@333fred 

The [NotNull] attribute seems to do the trick! I thought that was only used to signal that a method won't return null even if the return type is nullable, but it seems it also has a second function:

> Specifies that an input argument was not null when the call returns.

Which is exactly what I'm looking for. Thanks a lot!
